### ISSUE TITLE
Help needed for serialisation / deserialisation

### ISSUE THREAD

**Issue by rramillien** [cycfi/elements#268]
Hello everyone,

I can't find a simple way to (de)serialize elements.
My need is standard, I want to have an "editor" that serialize the UI to any format (probably JSON)
And I want to deserialize it in the host.

I tried using nlohmann/json, but I'm stuck.
From what I can see, I can only build elements when a child already exists. For e:xample (gotten from basicSlidersAndKnobs)
return
      margin({ 20, 10, 20, 10 },
         vmin_size(400,
            htile(
               margin({ 20, 20, 20, 20 }, pane("Vertical Sliders", make_vsliders(), 0.8f)),
               margin({ 20, 20, 20, 20 }, pane("Horizontal Sliders", make_hsliders(), 0.8f)),
               hstretch(0.5, margin({ 20, 20, 20, 20 }, pane("Knobs", make_dials(), 0.8f)))
            )
         )
      );

And classes seems to be some kind of "immutable" and do not allow to change children at runtime. Am I wrong ?
What I would have done is instanciate the parent, then add all its children while deserialising them.

Is it possible ? Or if not, does anybody already serialised/deserialised elements and have example or explanation ?

Best regards,

**Comment by djowel** [comment]
You should use escapes to format your code here. Example:

```
   struct foo
   {
      int bar;
   };
```

**Comment by djowel** [comment]
Anyway, you've barely touched the surface. The front-end data structure will be a major undertaking in and of itself, in the standpoint of design. As I said, it can't be too low-level. It is possible to map element's structures one-to-one, but then I'm afraid it will not be performant due to type erasure (assuming you know what that means).

But again, the real question is: why bother to write it in JSON if you can write more elegantly in c++? You still have not explained why designing the UI without code is mandatory for you. JSON is code. YAML is code.

**Comment by rramillien** [comment]
mmh, yes, effectively... you're right about type erasure. There is many years I had not really coded in C++ and I didn't think of this.... But I will make some tests and go back to you if I'm stuck.

About my need, it is simple. I want to create a open-source VST with is parameterable. Think of it as a simple reaktor or simple kontakt. So the user will be able to design its own elements and bind the controls to functions.
So because user have to design is own view, I need to serialize/deserialize it. I don't want the user to code and compile its own view... I want to make it simple for user. First step will be deserialization to build a view of elements from json.

**Comment by rramillien** [comment]
Mmmmh. Perhaps a good solution could be to create my own "standard" high level elements based on your generic ones ?
JSON will then be based on this "standard" non generic ones ? As you said, it's not too low level and it fulfill my requirements.

**Comment by rramillien** [comment]
That could lead to someting like this:

```
#pragma once

#include <elements.hpp>

#include "BasicTrack.hpp"

using namespace cycfi::elements;

#define SLIDER_MARKS_START namespace live::tritone::vie::ui::model { \
    struct SliderMarks {

#define SLIDER_MARKS_END std::string trackId; \
\
    using elementsMap = std::map<std::string const, void*>; \
    static auto deserialize(elementsMap& elements) { \
        auto track = BasicTrack::deserialize(elements); \
        return slider_marks<size>(track); \
    } \
}; \
}

#define SIZE_10 static const int size = 10;
#define SIZE_20 static const int size = 20;
#define SIZE_30 static const int size = 30;
#define SIZE_40 static const int size = 40;

#define SLIDER_10 SLIDER_MARKS_START SIZE_10 SLIDER_MARKS_END
#define SLIDER_20 SLIDER_MARKS_START SIZE_20 SLIDER_MARKS_END
#define SLIDER_30 SLIDER_MARKS_START SIZE_30 SLIDER_MARKS_END
#define SLIDER_40 SLIDER_MARKS_START SIZE_40 SLIDER_MARKS_END
```

**Comment by djowel** [comment]
Yes! That is certainly possible! Use `shared` and `element_ptr` to build your dynamic controls.

https://github.com/cycfi/elements/blob/master/lib/include/elements/element/element.hpp#L71-L81

**Comment by rramillien** [comment]
Thanks Joel,

Ok, I will try your solutions this week.
I will tell you if I need help.

Best regards,
